File: test/clean-main-worktree.test.ts
import { describe, expect, it, afterAll } from 'vitest';
import { execFileSync } from 'node:child_process';
import { existsSync, mkdtempSync, mkdirSync, readFileSync, rmSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { assertCleanMainWorktree, popStashBySha, stashMainWorktree } from '../src/git/worktree.js';

const dirs: string[] = [];
afterAll(() => {
  for (const d of dirs) rmSync(d, { recursive: true, force: true });
});

function initRepo(): string {
  const dir = mkdtempSync(join(tmpdir(), 'da-cleanmain-'));
  dirs.push(dir);
  const repo = join(dir, 'repo');
  mkdirSync(repo);
  execFileSync('git', ['init', '-b', 'main'], { cwd: repo });
  execFileSync('git', ['config', 'user.email', 'test@example.com'], { cwd: repo });
  execFileSync('git', ['config', 'user.name', 'test'], { cwd: repo });
  writeFileSync(join(repo, 'f.txt'), 'x\n');
  execFileSync('git', ['add', '.'], { cwd: repo });
  execFileSync('git', ['commit', '-m', 'init'], { cwd: repo });
  return repo;
}

function porcelain(repo: string): string {
  return execFileSync('git', ['status', '--porcelain'], { cwd: repo }).toString();
}

describe('stashMainWorktree', () => {
  it('returns null on a clean tree without creating a stash', async () => {
    const repo = initRepo();
    expect(await stashMainWorktree(repo, 'test')).toBeNull();
    expect(execFileSync('git', ['stash', 'list'], { cwd: repo }).toString()).toBe('');
  });

  it('stashes tracked and untracked changes and restores them on pop by SHA', async () => {
    const repo = initRepo();
    writeFileSync(join(repo, 'f.txt'), 'modified\n');
    writeFileSync(join(repo, 'untracked.txt'), 'new\n');
    expect(porcelain(repo)).not.toBe('');

    const sha = await stashMainWorktree(repo, 'devagent auto-stash before merge');
    expect(sha).toMatch(/^[0-9a-f]{40}$/);
    expect(porcelain(repo)).toBe('');

    expect(await popStashBySha(repo, sha!)).toBe(true);
    expect(readFileSync(join(repo, 'f.txt'), 'utf8')).toBe('modified\n');
    expect(readFileSync(join(repo, 'untracked.txt'), 'utf8')).toBe('new\n');
    // pop consumed the stash entry: no stash refs remain
    expect(execFileSync('git', ['stash', 'list'], { cwd: repo }).toString()).toBe('');
  });
});

describe('assertCleanMainWorktree', () => {
  it('resolves on a clean tree checked out on main', async () => {
    const repo = initRepo();
    await expect(assertCleanMainWorktree(repo, 'main')).resolves.toBeUndefined();
  });

  it('rejects with a diagnostic message on detached HEAD', async () => {
    const repo = initRepo();
    execFileSync('git', ['checkout', '--detach'], { cwd: repo });
    await expect(assertCleanMainWorktree(repo, 'main')).rejects.toThrow(
      'main worktree is in detached-HEAD state; refusing to merge',
    );
  });

  it('rejects with the branch name on a non-main branch', async () => {
    const repo = initRepo();
    execFileSync('git', ['checkout', '-b', 'feature'], { cwd: repo });
    await expect(assertCleanMainWorktree(repo, 'main')).rejects.toThrow(
      /main worktree is on branch feature, expected main; refusing to merge/,
    );
  });

  it('rejects with a porcelain preview when the tree is dirty', async () => {
    const repo = initRepo();
    writeFileSync(join(repo, 'f.txt'), 'dirty\n');
    await expect(assertCleanMainWorktree(repo, 'main')).rejects.toThrow(
      /main worktree has uncommitted changes; refusing to merge:\n\s*M f\.txt/,
    );
  });
});

File: internal/git/clean_main_worktree_test.go
package git

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// Ports test/clean-main-worktree.test.ts (vitest) — the stash-by-SHA and
// clean-main-worktree guard matrix (clean / dirty / detached / wrong-branch),
// coverage added with PR #84.

func porcelain(t *testing.T, repo string) string {
	t.Helper()
	return runGit(t, repo, "status", "--porcelain")
}

func TestStashMainWorktreeNullOnCleanTree(t *testing.T) {
	repo := initRepo(t)
	sha, err := StashMainWorktree(repo, "test")
	if err != nil {
		t.Fatal(err)
	}
	if sha != "" {
		t.Fatalf("sha = %q, want empty", sha)
	}
	if list := runGit(t, repo, "stash", "list"); list != "" {
		t.Fatalf("stash list = %q", list)
	}
}

func TestStashMainWorktreeStashAndPopBySha(t *testing.T) {
	repo := initRepo(t)
	writeFile(t, filepath.Join(repo, "f.txt"), "modified\n")
	writeFile(t, filepath.Join(repo, "untracked.txt"), "new\n")
	if porcelain(t, repo) == "" {
		t.Fatal("tree should be dirty")
	}

	sha, err := StashMainWorktree(repo, "devagent auto-stash before merge")
	if err != nil {
		t.Fatal(err)
	}
	if len(sha) != 40 {
		t.Fatalf("sha = %q", sha)
	}
	if porcelain(t, repo) != "" {
		t.Fatalf("tree should be clean after stash: %q", porcelain(t, repo))
	}

	if !PopStashBySha(repo, sha) {
		t.Fatal("pop should succeed")
	}
	if data, err := os.ReadFile(filepath.Join(repo, "f.txt")); err != nil || string(data) != "modified\n" {
		t.Fatalf("f.txt = %q, %v", data, err)
	}
	if data, err := os.ReadFile(filepath.Join(repo, "untracked.txt")); err != nil || string(data) != "new\n" {
		t.Fatalf("untracked.txt = %q, %v", data, err)
	}
	// pop consumed the stash entry: no stash refs remain
	if list := runGit(t, repo, "stash", "list"); list != "" {
		t.Fatalf("stash list = %q", list)
	}
}

func TestPopStashByShaFalseOnConflictLeavesStashIntact(t *testing.T) {
	repo := initRepo(t)
	writeFile(t, filepath.Join(repo, "f.txt"), "stashed version\n")
	sha, err := StashMainWorktree(repo, "will conflict")
	if err != nil || sha == "" {
		t.Fatalf("stash = %q, %v", sha, err)
	}
	// Conflicting uncommitted change: apply fails, stash must survive.
	writeFile(t, filepath.Join(repo, "f.txt"), "divergent\n")
	if PopStashBySha(repo, sha) {
		t.Fatal("pop should report failure on conflict")
	}
	if list := runGit(t, repo, "stash", "list", "--format=%H"); !strings.Contains(list, sha) {
		t.Fatalf("stash must remain intact, list = %q", list)
	}
}

func TestAssertCleanMainWorktreeResolvesOnCleanMain(t *testing.T) {
	repo := initRepo(t)
	if err := AssertCleanMainWorktree(repo, "main"); err != nil {
		t.Fatalf("clean main should pass: %v", err)
	}
}

func TestAssertCleanMainWorktreeRejectsDetachedHead(t *testing.T) {
	repo := initRepo(t)
	runGit(t, repo, "checkout", "--detach")
	err := AssertCleanMainWorktree(repo, "main")
	if err == nil || err.Error() != "main worktree is in detached-HEAD state; refusing to merge" {
		t.Fatalf("err = %v", err)
	}
}

func TestAssertCleanMainWorktreeRejectsWrongBranch(t *testing.T) {
	repo := initRepo(t)
	runGit(t, repo, "checkout", "-b", "feature")
	err := AssertCleanMainWorktree(repo, "main")
	want := "main worktree is on branch feature, expected main; refusing to merge"
	if err == nil || err.Error() != want {
		t.Fatalf("err = %v", err)
	}
}

func TestAssertCleanMainWorktreeRejectsDirtyTreeWithPreview(t *testing.T) {
	repo := initRepo(t)
	writeFile(t, filepath.Join(repo, "f.txt"), "dirty\n")
	err := AssertCleanMainWorktree(repo, "main")
	if err == nil {
		t.Fatal("dirty tree must be rejected")
	}
	if !strings.HasPrefix(err.Error(), "main worktree has uncommitted changes; refusing to merge:\n") {
		t.Fatalf("err = %q", err.Error())
	}
	if !strings.Contains(err.Error(), "M f.txt") {
		t.Fatalf("porcelain preview missing: %q", err.Error())
	}
}

// ensureStateBranch fixture parity: the state-branch tests live in
// state_branch_test.go; the shared fixture lives in helpers_test.go.
var _ = filepath.Join
